[Perl] Adding Gluon interface to Perl, miscellaneous changes in order to sync with Python

sergeykolychev opened the pull request:
@tlby could you please see if you can spot anything out of place ? Thanks.

sergeykolychev commented:
@piiswrong please take a look when you have time, thank you.

piiswrong commented:
As I can't read perl and this is a good patch, it would be good if you get get someone familiar with perl to take a look

sergeykolychev commented:
@piiswrong I asked another Perl contributor, @tlby to review

tlby commented:
> @tlby could you please see if you can spot anything out of place ? Thanks.

@sergeykolychev, certainly.  I'd like to check it out and get my hands dirty though, so it might be a day or two to get up to speed on the work here.

sergeykolychev commented:
@tlby sure, take as long as you need.

tlby commented on perl-package/AI-MXNet/lib/AI/MXNet/Logging.pm:
```diff
@@ -17,8 +17,11 @@
+use warnings;
 use Mouse;
-sub warning { shift; warn sprintf(shift, @_) . "\n" };
+$AI::MXNet::Logging::silent = 0;
```
`our $silent = 0;` would achieve the same thing here, and then uses of the variable in this package wouldn't need to be fully qualified.

sergeykolychev replied:
ok, I'll change, thanks.

tlby commented on perl-package/AI-MXNet/lib/AI/MXNet/Gluon/NN/ConvLayers.pm:
```diff
@@ -0,0 +1,1363 @@
+method _alias() { 'conv' }
+
+use Data::Dumper;
+use overload '""' => sub {
```
translating `def __repr__` to `use overload '""' => sub { ... }` may cause some collisions if the python guys add a `def __str__` later on or if the algorithms expects str(...) to be more performant than the `__repr__` implementation.  an explicit object method might make sense, but obviously it would be a hassle to track down all the places we'd need to invoke it where python was calling it implicitly.

sergeykolychev replied:
yes, let's worry about this later.

sergeykolychev replied:
but we need bool overload for these objects that's for sure. I'll add that later, or may be update mouse to have global bool overload aliased to defined.

tlby commented on perl-package/AI-MXNet/lib/AI/MXNet/Metric.pm:
```diff
@@ -20,6 +20,7 @@ use strict;
 use warnings;
 use AI::MXNet::Function::Parameters;
 use Scalar::Util qw/blessed/;
+use JSON::PP;
```
I think normally people just use `JSON` and that module makes a good choice about using `JSON::PP` or `JSON::XS`.

sergeykolychev replied:
JSON::PP is in perl core, JSON or JSON::XS are not, this had motivated my decision.

tlby replied:
ah, news to me. thanks!

tlby commented on perl-package/AI-MXNet/lib/AI/MXNet/Symbol/NameManager.pm:
```diff
@@ -75,10 +75,10 @@ method get(Maybe[Str] $name, Str $hint)
 }
 
-$AI::MXNet::current_nm_ldr = __PACKAGE__->new;
+$mx::NameManager = __PACKAGE__->new;
```
this variable would be better inside this module's namespace.

sergeykolychev replied:
This var is sometimes exposed to user asking them to alter default naming pattern via
local($mx::NameManager) = mx->name->Prefix("prefix_");
This is consistent with local($mx::AttrScope) = mx->AttrScope(...);
and local($mx::Context) = mx->gpu(1); 
which is in its turn consistent with the python examples.

sergeykolychev replied:
Actually you're right I'll fix that particular usage, good catch.

sergeykolychev commented:
@piiswrong 
The code has been reviewed and approved by another perl contributor @tlby; I have addressed the concerns that he raised and also incorporated recent changes on python side such as CudaModule and random functions redesign. The perl cpu/gpu tests passed. Could you please merge it at your convenience, thanks.

sergeykolychev commented:
@piiswrong just making sure that previous comment is not lost in your mail.

piiswrong commented:
Thanks!

gautamkmr commented:
Why did we merge since jenkins failed on this PR ?
There has been recent build  481, 483, 490 which are failing.

sergeykolychev commented:
@gautamkmr @indhub 
This just one test that depends on the results of training, Python test had 200 epochs for training I felt that it's too much , but it looks rarely it's needed I added pull to increase the number of epochs for test_loss.t this should close the matter please always contact me immediately if Perl test fails.
